server/middleware.ts

```diff
@@ -72,11 +76,25 @@ export function refreshSession() {
     const session = getSession<SessionData, FlashData>(c);
     const auth = session.get(authSessionKey);
 
-    if (!auth || !isExpiringSoon(auth.expiresAt)) {
+    if (!auth) {
       return next();
     }
-
+    let isValidSession = true;
+    if (!isExpiringSoon(auth.expiresAt)) {
+      isValidSession = await validateSession(auth.refreshToken);
+      if (isValidSession) {
```

Comment:
Yeah indeed that makes more sense.
@rajeshj11 I am not even sure whats the purpose of those changes? Aren't those left over from trying the different approach. In main we handle it like this:
```tsx
/**
 * Refresh access token middleware
 *
 */
export function refreshSession() {
  return createMiddleware(async (c, next) => {
    //@ts-expect-error fixed soon
    const session = getSession<SessionData, FlashData>(c);
    const auth = session.get(authSessionKey);

    if (!auth || !isExpiringSoon(auth.expiresAt)) {
      return next();
    }

    try {
      session.set(authSessionKey, await refreshAccessToken(auth.refreshToken));
    } catch (cause) {
      session.flash(
        "errorMessage",
        "You have been logged out. Please log in again."
      );

      session.unset(authSessionKey);
    }

    return next();
  });
}

```

Reply:
> I would go `next` if `isValidSession` && `!isExpiringSoon`. Else, we can continue and try to `refreshAccesToken`. It should failed for invalid sessions since the refresh token has been revoked on password change.

The above logic ensures that validateSession is only triggered when the local session does not expire soon. if it is expiring soon it will trigger the direct refreshAccestoken(let's say if the password changed. it will still throw error while trying to refresh token using refresh token)
FYI: @DonKoko

Reply:
it was re-written like this

``` if (!auth) {//if no auth
      return next();
    }
    let isValidSession = true;
    if (!isExpiringSoon(auth.expiresAt)) {// if not expiring soon then only hit the validateSession
      isValidSession = await validateSession(auth.refreshToken);
      if (isValidSession) {// if valid and and not expiring soon we will go to next
        return next();
      }
    }
    try {
      //if not true. then we can throw error
      if (!isValidSession) {// if not a valid session we simply throw an error which will clear the session.
        throw new ShelfError({
          cause: null,
          message: "Your session has expired. Please log in again.",
          label: "Auth",
        });
      }
     //only hit when the toke is expiring soon
      session.set(authSessionKey, await refreshAccessToken(auth.refreshToken));
    } catch (cause) {
      session.flash(
        "errorMessage",
        "You have been logged out. Please log in again."
      );

      session.unset(authSessionKey);
    }```

Reply:
Ok I can understand the confusion here. Maybe the 'isValidSession' code should be moved to the `protect` middleware and `refreshSession` reverted?

Protect purpose is to check that a given route is protected (😅).
Sorry for not having seen this before @DonKoko

Reply:
got it

Reply:
@rphlmr this has been updated. Can you take a look and lmk what you think?